src/basic/fs-util.c

```diff
@@ -1028,3 +1029,44 @@ int parse_cifs_service(
 
         return 0;
 }
+
+char *resolve_auto_dir(const char *path) {
+        _cleanup_free_ char *parent = NULL;
+        _cleanup_free_ char *glob_path = NULL;
+        _cleanup_globfree_ glob_t pglob = {};
+        int r;
+
+        if (!path)
+                return NULL;
+
+        log_debug("Trying to version %s.", path);
+        r = path_extract_directory(path, &parent);
+        if (r < 0 || !endswith(parent, "auto.d"))
+                goto error;
+
+        r = asprintf(&glob_path, "%s*", path);
+        if (r < 0)
+                goto error;
+
+        r = safe_glob(glob_path, 0, &pglob);
```

Comment:
please don't use globs for this. globs are OK if users actually provide us with a glob expression. But if we use a constant our even generated glob expression just use opendir()/readdir() insteads, it's simpler and faster.

Reply:
also, I think we should skip any dot prefixed entries, and filter out entries of type we are not interested in (i.e. if we are called for S_IFDIR ignore anything that is not S_IFDIR, and if we are called for S_IFREG similar). (But mybe allow symlinks in either case)

Reply:
Finally, I'd leave a door open for later extensions: for sd-boot we reserve entry names starting in a `@` to have special meanings: we don't allow them to appear on disk, but they are resolved to special meaning in different contexts. I would keep the door open. Hence please ignore any entries starting with `@` and debug log about them, that they are reserved (to make this debuggable). That way we can be sure not to create compat problems should we later on give them meaning.

(what do i mean by all this: specifically, we could eventually allow users to define a symlink `@default` that can point to any entry, and that allows users to override the version picked entry and make some explicitly picked version the one to use).

e.g:

```
foo.auto.d/0.1/
foo.auto.d/0.2/
foo.auto.d/0.3/
foo.auto.d/0.4/
foo.auto.d/@default → 0.3
```

in such a setup 0.4 would of course be the entry to boot if nothing is specified. But if the `@default` symlink exists we'd allow users to override this behaviour that way, and (temporarily?) pick a different one (i.e. 0.3 in this case)

Now, I am not saying we should implement this immediately (though if you like, go ahead, I think it makes sense), but I'd definitely want to keep the door open for this, and hence for now ignore all `@` prefixed entries.

Reply:
Makes sense

Reply:
Last version doesn't use glob() but readdir(). I've added filename checks including `@`.